aws-rds-cfn-common/src/main/java/software/amazon/rds/common/util/DriftDetector.java

```diff
@@ -0,0 +1,252 @@
+        this.schema = schema;
+        this.writeOnlyLookup = new HashSet<>(
+                Optional.ofNullable(schema.getWriteOnlyPropertiesAsStrings())
+                        .orElse(Collections.emptyList())
+        );
+    }
+
+    private static boolean isTransformableProperty(
+            final ResourceTypeSchema schema,
+            final String propertyName
+    ) throws DriftDetectorRuntimeException {
+        return schema.getPropertyTransform().containsKey(propertyName);
+    }
+
+    private static Object castToPrimitive(final JsonNode node, final Class<?> klass) {
+        if (Integer.class.equals(klass)) {
+            return node.asInt();
+        } else if (Boolean.class.equals(klass)) {
+            return node.asBoolean();
+        } else if (String.class.equals(klass)) {
+            return node.asText();
+        }
+
+        throw new DriftDetectorRuntimeException("Unexpected cast input class: " + klass.getName());
+    }
+
+    private List<Object> getTransformations(
+            final Object in,
+            final String path,
+            final JsonNode rootNode
+    ) {
+        final List<Object> result = new ArrayList<>();
+        final String propTxs = schema.getPropertyTransform().get(path);
+        if (propTxs != null) {
+            for (final String tx : propTxs.split(PROPERTY_OR_SPLIT_REGEX)) {
+                try {
+                    final Expressions expr = Expressions.parse(tx);
+                    final JsonNode txNode = expr.evaluate(rootNode);
+                    result.add(castToPrimitive(txNode, in.getClass()));
+                } catch (ParseException | IOException | EvaluateException e) {
+                    throw new DriftDetectorRuntimeException("Failed to parse jsonata expression: " + tx, e);
+                }
+            }
+        }
+        return result;
+    }
+
+    private boolean isWriteOnly(final String path) {
+        return writeOnlyLookup.contains(path);
+    }
+
+    public <T> Map<String, Mutation> detectDrift(final T prev, final T upd) {
+        try {
+            final JsonNode rootNode = objectMapper.readTree(objectMapper.writeValueAsString(prev));
+            return detectDrift(prev, upd, PROPERTIES_ROOT, rootNode);
+        } catch (JsonProcessingException e) {
+            throw new DriftDetectorRuntimeException("Failed to parse the root node", e);
+        }
+    }
+
+    private <T> Map<String, Mutation> detectDriftTransformable(
+            final T prev,
+            final T upd,
+            final String path,
+            final JsonNode rootNode
+    ) {
+        final List<Object> transformations = getTransformations(prev, path, rootNode);
+
+        if (transformations.stream().anyMatch(tx -> {
+            if (Objects.equals(tx, upd)) {
+                return true;
+            }
+            if (prev instanceof String) {
+                final String transformed = tx.toString().replaceAll("^\"|\"$", "");
```

Comment:
This one strips double quotes from string literals that would be returned by JSONata: say. It always returns a valid JSON token. In case of a string, it would return "\"foo\"" instead of "foo".